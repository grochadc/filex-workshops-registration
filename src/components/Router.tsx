import React, { useState } from "react";

type RouterContextType = {
  contextRoute: string;
  setRoute: (route: string) => void;
};

function initFn(route) {
  return true;
}
const initialContext = {
  contextRoute: "",
  setRoute: initFn,
};

const RouterContext = React.createContext<RouterContextType>(initialContext);

type RouterProps = {
  children: React.ReactNode;
  defaultRoute: string;
};

const Router = ({ defaultRoute, children }: RouterProps) => {
  const [route, setRoute] = useState(defaultRoute);
  return (
    <RouterContext.Provider value={{ contextRoute: route, setRoute }}>
      {children}
    </RouterContext.Provider>
  );
};

type RouterViewProps = {
  route: string;
  children: React.ReactNode;
};
const useRoute = () => React.useContext(RouterContext);

const View = ({ route, children }: RouterViewProps) => {
  const { contextRoute } = useRoute();
  return <>{route === contextRoute ? children : null}</>;
};

Router.View = View;
Router.useRoute = useRoute;

export default Router;
