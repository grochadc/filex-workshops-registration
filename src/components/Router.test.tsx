import React from "react";
import { render } from "@testing-library/react";
import Router from "./Router";

const App = () => {
  return (
    <Router defaultRoute={"home"}>
      <Router.View route="home">This is the homepage.</Router.View>
      <Router.View route="selection">This is the selection page.</Router.View>
    </Router>
  );
};

test("Renders without crashing [Snapshot]", () => {
  const { asFragment } = render(<App />);
  expect(asFragment()).toMatchSnapshot();
});

test("Renders only one route and nothing else", () => {
  const { getByText, queryByText } = render(<App />);
  expect(getByText(/homepage/i)).toBeInTheDocument();
  expect(queryByText(/selection/i)).not.toBeInTheDocument();
});
