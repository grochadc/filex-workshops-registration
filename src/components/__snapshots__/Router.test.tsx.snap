// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without crashing [Snapshot] 1`] = `
<DocumentFragment>
  This is the homepage.
</DocumentFragment>
`;
